messaging/src/main/java/org/axonframework/common/infra/JacksonComponentDescriptor.java

```diff
@@ -0,0 +1,190 @@
+            var componentSeenAlready = id != null;
+            if (componentSeenAlready) {
+                var refNode = objectMapper.createObjectNode();
+                refNode.put("$ref", id);
+                describeType(component, refNode);
+                return refNode;
+            }
+            return describeComponentJson(component);
+        }
+        return objectMapper.valueToTree(object);
+    }
+
+    private ObjectNode describeComponentJson(DescribableComponent component) {
+        var componentId = System.identityHashCode(component) + "";
+
+        // Register this component before processing its properties.
+        // This prevents infinite recursion with circular references.
+        processedComponents.put(component, componentId);
+
+        var nestedDescriptor = new JacksonComponentDescriptor(this.objectMapper, this.processedComponents);
+        describeIdAndType(component, nestedDescriptor);
+        component.describeTo(nestedDescriptor);
+        return nestedDescriptor.rootNode;
+    }
+
+    private static void describeIdAndType(
+            DescribableComponent component,
+            JacksonComponentDescriptor componentDescriptor
+    ) {
+        componentDescriptor.rootNode.put("_id", System.identityHashCode(component) + "");
+        describeType(component, componentDescriptor.rootNode);
+    }
+
+    private static void describeType(DescribableComponent component, ObjectNode objectNode) {
+        objectNode.put("_type", component.getClass().getSimpleName());
+    }
+
+    @Override
+    public void describeProperty(@Nonnull String name, @Nonnull Collection<?> collection) {
+        var arrayNode = objectMapper.createArrayNode();
+        for (var item : collection) {
+            var json = describeObject(item);
+            arrayNode.add(json);
+        }
+        rootNode.set(name, arrayNode);
+    }
+
+    @Override
+    public void describeProperty(@Nonnull String name, @Nonnull Map<?, ?> map) {
+        ObjectNode mapNode = objectMapper.createObjectNode();
+
+        for (var entry : map.entrySet()) {
+            var key = entry.getKey().toString();
+            var value = entry.getValue();
+            var json = describeObject(value);
+            mapNode.set(key, json);
+        }
+
+        rootNode.set(name, mapNode);
+    }
+
+    @Override
+    public void describeProperty(@Nonnull String name, @Nonnull String value) {
+        rootNode.put(name, value);
+    }
+
+    @Override
+    public void describeProperty(@Nonnull String name, long value) {
+        rootNode.put(name, value);
+    }
+
+    @Override
+    public void describeProperty(@Nonnull String name, boolean value) {
+        rootNode.put(name, value);
+    }
+
+    @Override
+    public String describe() {
+        try {
+            return objectMapper.writeValueAsString(rootNode);
+        } catch (Exception e) {
+            throw new RuntimeException("Error generating JSON description: " + e.getMessage(), e);
```

Comment:
Nit: Maybe it would make sense to have a specific `ComponentDescriptorException`, assuming there will be more outliers we need to deal with. Wdyt?

Reply:
Added.